### DISCUSSION TITLE
최근 javascript의 동향에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by airman5573** [woowacourse-study/2022-woowahan-bansanghwe#125] [JavaScript]
## ES2015

ES2015에 다음과 같은 기능들이 ECAMScript Spec에 추가되었습니다.  
[arrows](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Functions/Arrow_functions)
[classes](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Classes)
[enhanced object literals](https://blog.webdevsimplified.com/2021-02/javascript-enhanced-object-literals/)
[template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)
[destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)
[default + rest + spread](https://medium.com/ecmascript-2015/default-rest-spread-f3ab0d2e0a5e)
[let](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Statements/let) + [const](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Statements/const)
[for .. of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)
[iterators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol)
[generators](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Generator)
[modules](https://developer.mozilla.org/ko/docs/Web/JavaScript/Guide/Modules)
[module loaders](https://github.com/lukehoban/es6features#module-loaders)
[Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)
[Set](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Set)
[WeakMap](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/WeakMap)
[WeakSet](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/WeakSet)
[proxies](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Proxy)
[symbols](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Symbol)
[subclassable built-ins](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Symbol)
[Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)
[math + number + string + array + object APIs](https://github.com/lukehoban/es6features#math--number--string--array--object-apis)
[reflect api](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Reflect)
[tail calls](https://github.com/lukehoban/es6features#tail-calls)

화살표 함수, 클래스, 모듈, promise 등등 정말 유용한 기능들이 많이추가 되었습니다.  

대부분의 브라우저가 ES2015는 이제 기본으로 지원합니다.  

## ES2016
Array.prototype.includes() 메서드가 추가된것 이외에 눈에 띄는 변화는 없습니다.

## ES2017
ES2017에는 다음과 같은 기능들이 추가되었습니다.  

[Object.values()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values)
[Object.entries()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries)
[String.prototype.padStart()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart)
[String.prototype.padEnd()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padEnd)
[Trailing comma](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas)
[Async/Await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await)
[Rest](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters)/[Spread](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) Properties
[Object.getOwnPropertyDescriptors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptors)

저희가 자주 사용하는 Async/Await와 Rest/Spread Properties 기능이 이때 추가 되었습니다.  

## ES2018
ES2018에는 다음과 같은 기능들이 추가되었습니다.  

[SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer)
[Atomics](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics)
[Tagged Template](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates)
[Object rest and spread properties](https://v8.dev/features/object-rest-spread)
[Promise.prototype.finally()](https://v8.dev/features/promise-finally)

SharedArrayBuffer, Atomics는 거의 저희가 쓸일은 없지만 앞으로 멀티쓰레딩 프로그래밍을 할 때 유용하게 사용될 수 있습니다.   
쓰레드끼리 메모리를 공유할 때 쓰입니다.  

Tagged Template도 유용한데요, 저희가 자주 사용하는 Emotion, StyledComponent에서 사용합니다.

Promise.prototype.finally() 도 요긴하게 사용될 수 있습니다. 이래나저래나 호출하겠다는 의미입니다.

## ES2019
배열에는 다음과 같은 기능들이 추가되었습니다.  
[Array.prototype.{flat,flatMap](https://exploringjs.com/es2018-es2019/ch_array-prototype-flat-flatmap.html)[}](https://exploringjs.com/es2018-es2019/ch_array-prototype-flat-flatmap.html)
[Array.prototype.flat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat)
[Array.prototype.flatMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap)
[Object.fromEntries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries)
[String.prototype.trimStart()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trimStart)
[String.prototype.trimEnd()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trimStart)
[Optional catch binding](https://exploringjs.com/es2018-es2019/ch_optional-catch-binding.html)

배열에 요긴한 함수가 생겼습니다. flat, flatMap !  원소들을 평평하게 만들 때 사용합니다.  

Optional catch binding 덕분에 catch 블럭에서 error argument를 받지 않아도 됩니다.  
소소하지만 편한 기능입니다.  

## ES2020
[Dynamic import](https://exploringjs.com/impatient-js/ch_modules.html#dynamic-imports)
[Namespace re-exporting](https://exploringjs.com/impatient-js/ch_modules.html#module-exports)
[Optional Chaining](https://exploringjs.com/impatient-js/ch_objects.html#optional-chaining)
[Nullish coalescing operator(??)](http://nullish%20coalescing%20operator%20%28/??))
[String.prototype.matchAll()](https://exploringjs.com/impatient-js/ch_regexps.html#String.prototype.matchAll)
[Promise.allSettled](https://exploringjs.com/impatient-js/ch_promises.html#Promise.allSettled)
[globalThis](https://exploringjs.com/impatient-js/ch_variables-assignment.html#globalThis)

Dynamic import 가 추가되었습니다. 덕분에 런타임에서 비동기적으로 js를 불러오고 실행할 수 있게되었습니다.  

Optional Chaning, Nullish coalescing 이 나왔습니다. 매우 편한 기능들입니다!  

Promise에 allSettled 기능이 나왔습니다.  하나라도 실패했을때 "실패"로 판정하고 싶다면 all을 사용하고, 몇개는 성공하고 몇개는 실패해도 되는 경우에는 allSettled를 사용합니다.  
저라면,,,,어지간 하면 allSettled를 all보다 선호할 듯 싶습니다!.

**Comment by al-bur** [comment]
wow

**Comment by byhhh2** [comment]
탐구하는 개발자는 무엇을 뜻하나요?

**Reply by byhhh2** [reply]
혼자 탐구하시나요?

**Reply by byhhh2** [reply]
탐구 경험에서 뭘 얻으셨나요?

**Reply by byhhh2** [reply]
라이브러리 까본 이유

**Reply by byhhh2** [reply]
왜 hook-form을 깠죠?

**Comment by byhhh2** [comment]
react에서 form을 다루는 방법?

**Reply by byhhh2** [reply]
제어 vs 비제어 어떤 방법 선호하시나요?

**Comment by byhhh2** [comment]
렌더링 최적화 경험

**Reply by byhhh2** [reply]
useCallback을 쓰는 이유

**Comment by byhhh2** [comment]
context API의 단점을 해결하는 방법

**Comment by JUDONGHYEOK** [comment]
팀프로젝트를 하면서 어려웠던점?

**Reply by JUDONGHYEOK** [reply]
컨벤션 맞추는게 어려웠던 이유?

**Reply by JUDONGHYEOK** [reply]
의견 충돌이 있을 때 해결하는 법

**Reply by byhhh2** [reply]
내 의견이 진짜 맞다고 생각하는데 받아드려지지 않을때 어떻게 하나요?

**Comment by JUDONGHYEOK** [comment]
마감기한이 있는 프로젝트에서 기능구현, 클린코드 선택

**Comment by JUDONGHYEOK** [comment]
좋은 팀이란 무엇인가?

**Reply by JUDONGHYEOK** [reply]
팀프로젝트에서 반대를 많이 나오게 한 노력

**Comment by JUDONGHYEOK** [comment]
병민은 어떤 개발자가 되고싶은지?

**Comment by byhhh2** [comment]
최근 프로젝트를 진행하면서 겪은 가장 큰 어려움?

**Comment by byhhh2** [comment]
왜 주변에 재미를 주는 개발자가 되고 싶나요?

**Comment by byhhh2** [comment]
최근에도 컨벤션을 맞추고 계신다고 하는데 전보다 나아진 점이 확실히 있나요?

**Comment by byhhh2** [comment]
자바스크립트에서 어려운 개념은?

**Reply by byhhh2** [reply]
그것을 해결한 경험
